crates/runtime-impl/src/runtime.rs

```diff
@@ -114,17 +117,27 @@ fn compute_moho_state_commitment(_state: &MohoState) -> MohoStateCommitment {
 
 /// Computes the exported Moho state from the inner state, also checking the
 /// verification key and export correctness.
-fn compute_wrapping_moho_state<P: MohoProgram>(state: &P::State) -> MohoState {
-    let inner_root = P::compute_state_commitment(state);
-
-    let next_vk = P::extract_next_vk(state);
-
-    let export_state = P::extract_export_state(state);
-    if !check_export_state_structure(&export_state) {
+fn update_wrapping_moho_state<P: MohoProgram>(
+    moho_state: &mut MohoState,
+    step_output: &P::StepOutput,
+) {
+    let inner_state = P::extract_post_state(step_output);
+    let inner_root = P::compute_state_commitment(inner_state);
+
+    // Determine the next inner verification key: use the updated key if available, otherwise fall
+    // back to the previous one
+    let next_vk = match P::extract_next_vk(step_output) {
+        Some(vk) => vk,
+        None => moho_state.next_vk().clone(),
+    };
+
+    P::update_export_state(moho_state.export_state_mut(), step_output);
+    if !check_export_state_structure(moho_state.export_state()) {
         panic!("runtime: invalid export state structure");
     }
 
-    MohoState::new(inner_root, next_vk, export_state)
+    moho_state.update_next_vk(next_vk);
+    moho_state.update_inner_state(inner_root);
```

Comment:
Shouldn't modify this in-place, if possible.  We should construct a new `MohoState` with the new values by taking apart the one passed in.  It's more functional.